### PR TITLE
Aave protection connector

### PR DESCRIPTION
**Pull request by Gauddel**
Gelato Aave protection connector

### CODE REVIEW

File: contracts/polygon/connectors/gelato/aave-services/protection/main.sol [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    /// @dev Function for cancelling and removing allowance
+    /// of aToken to _protectionAction
+    function cancelAndRevoke() external payable {
+        if (_dsaHasProtection()) _cancelProtection();
+        _revokeAllowance();
+    }
+
+    function _submitProtection(
```
Comment (KaymasJain) [inline contracts/polygon/connectors/gelato/aave-services/protection/main.sol]: Can you move all the internal functions to `helpers.sol`? `main.sol` are mainly for functions that are going to be called by DSAs.

File: contracts/polygon/connectors/gelato/aave-services/protection/main.sol [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    /// @param _isPermanent boolean to set a protection as permanent
+    function submitProtection(
+        uint256 _wantedHealthFactor,
+        uint256 _minimumHealthFactor,
+        bool _isPermanent
+    ) external payable {
+        _submitProtection(
+            _wantedHealthFactor,
+            _minimumHealthFactor,
+            _isPermanent
```
Comment (KaymasJain) [inline contracts/polygon/connectors/gelato/aave-services/protection/main.sol:26]: Can we add events for these 4 main functions? Need to create an `events.sol` for it

File: contracts/polygon/connectors/gelato/aave-services/protection/main.sol (hunk 1 above)
Comment (Gauddel) [reply contracts/polygon/connectors/gelato/aave-services/protection/main.sol]: Yes all internals function from ```main.sol``` has been moved to ```helpers.sol```.

File: contracts/polygon/connectors/gelato/aave-services/protection/main.sol (hunk 2 above)
Comment (Gauddel) [reply contracts/polygon/connectors/gelato/aave-services/protection/main.sol:26]: Yes i added 3 events ```LogSubmitProtection```, ```LogUpdateProtection``` and ```LogCancelProtection```, theses events will be emitted inside these 4 externals functions.

Comment (Gauddel) [reply contracts/polygon/connectors/gelato/aave-services/protection/main.sol:26]: Do you want another event for revoke allowance, maybe like ```event LogRevokeAllowance(address dsa, address action)```?

File: contracts/polygon/connectors/gelato/aave-services/protection/events.sol [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    );
+    event LogUpdateProtection(
+        address indexed dsa,
+        address indexed action,
+        uint256 wantedHealthFactor,
+        uint256 minimumHealthFactor,
+        bool isPermanent
+    );
+    event LogCancelProtection(address indexed dsa, address indexed action);
+}
```
Comment (KaymasJain) [inline contracts/polygon/connectors/gelato/aave-services/protection/events.sol:21]: Can we also add an event like: `event LogCancelAndRevoke(address indexed dsa, address indexed action);`

File: contracts/polygon/connectors/gelato/aave-services/protection/main.sol [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        _cancelProtection();
+        emit LogCancelProtection(address(this), _protectionAction);
+    }
+
+    /// @dev Function for cancelling and removing allowance
+    /// of aToken to _protectionAction
+    function cancelAndRevoke() external payable {
+        if (_dsaHasProtection()) {
+            _cancelProtection();
+            emit LogCancelProtection(address(this), _protectionAction);
```
Comment (KaymasJain) [inline contracts/polygon/connectors/gelato/aave-services/protection/main.sol]: update this with `cancelAndRevoke` - https://github.com/Instadapp/dsa-connectors/pull/67/files#r693339574

File: contracts/polygon/connectors/gelato/aave-services/protection/events.sol (hunk 3 above)
Comment (Gauddel) [reply contracts/polygon/connectors/gelato/aave-services/protection/events.sol:21]: done

File: contracts/polygon/connectors/gelato/aave-services/protection/main.sol (hunk 4 above)
Comment (Gauddel) [reply contracts/polygon/connectors/gelato/aave-services/protection/main.sol]: done

**Review by KaymasJain — APPROVED** [review]

**Comment by KaymasJain** [comment]
Aave Polygon Protection Connector

ConnectorName: "GELATO-AAVE-A"
ConnectorAddress: [0x177Bd89A1D8643C9525D2DF131C1a6C513869299](https://polygonscan.io/address/0x177Bd89A1D8643C9525D2DF131C1a6C513869299)

**Comment by gitpusha** [comment]
Let us know if you need any help for the UI integration. 

You can find a lot of helper functions for submitting protections to the on-chain system and querying protection data from the subgraphs in our `aave-protection-lib` : https://github.com/gelatodigital/aave-protection-lib/blob/master/src/index.ts

https://www.npmjs.com/package/@gelatonetwork/aave-protection-lib

**Comment by KaymasJain** [comment]
Aave mainnet Protection connector - 

ConnectorName: "GELATO-AAVE-A"
ConnectorAddress: [0x825832a5A589ed9500CaEE2aa2D2c3117218D6A9](https://etherscan.io/address/0x825832a5A589ed9500CaEE2aa2D2c3117218D6A9)

File: docs/connectors.json [hunk 5]
```diff
@@ -24,15 +24,17 @@
       "WETH-A": "0x22075fa719eFb02Ca3cF298AFa9C974B7465E5D3",
       "REFINANCE-A": "0x6f22931423e8ffC8d51f6E5aF73118fC64b27856",
       "INST-A": "0x52C2C4a0db049255fF345EB9D3Fb1f555b7a924A",
-      "REFLEXER-A": "0xaC6dc28a6251F49Bbe5755E630107Dccde9ae2C8"
+      "REFLEXER-A": "0xaC6dc28a6251F49Bbe5755E630107Dccde9ae2C8",
+      "AAVE-PROT-A": "0x825832a5A589ed9500CaEE2aa2D2c3117218D6A9"
```
Comment (thrilok209) [inline docs/connectors.json]:
```suggestion
      "GELATO-AAVE-A: "0x825832a5A589ed9500CaEE2aa2D2c3117218D6A9"
```

File: docs/connectors.json [hunk 6]
```diff
@@ -24,15 +24,17 @@
...
     },
     "137" : {
       "1INCH-A": "0xC0d9210496afE9763F5d8cEb8deFfBa817232A9e",
       "AAVE-V2-A": "0xE84d8010Afc3663919F44685cB53ED88866da3eE",
       "AUTHORITY-A": "0xf73C94402BC24148b744083eD02654EEc2C37D5B",
       "BASIC-A": "0x1cAF5EC802ca602E98139AD96A8f2B7BC524264E",
       "AAVE-CLAIM-A": "0xC7Cb1dE2721BFC0E0DA1b9D526bCdC54eF1C0eFC",
-      "PARASWAP-A": "0xFb3a1D56eD56F046721B9aCa749895100754578b"
+      "PARASWAP-A": "0xFb3a1D56eD56F046721B9aCa749895100754578b",
+      "AAVE-PROT-A": "0x177Bd89A1D8643C9525D2DF131C1a6C513869299"
```
Comment (thrilok209) [inline docs/connectors.json]:
```suggestion
      "GELATO-AAAVE-A": "0x177Bd89A1D8643C9525D2DF131C1a6C513869299"
```

**Comment by thrilok209** [comment]
Mainnet connector whitelisted: [0xabc3d397f99ff6de1874d50ae916db8df6d95becc0be9ced0b4baa8d9859e0fe](https://etherscan.io/tx/0xabc3d397f99ff6de1874d50ae916db8df6d95becc0be9ced0b4baa8d9859e0fe)

Polygon connector whitelisted: [0x567eb8a3ad4872256e579e3662338bff8c7a1ab8556ff3dd6bf157320dd304fc](https://polygonscan.com/tx/0x567eb8a3ad4872256e579e3662338bff8c7a1ab8556ff3dd6bf157320dd304fc)

**Pull request merged**